### DISCUSSION TITLE
How to retain set parameters of device instead of returning to default values (driver specified inputs) after restart?

### DISCUSSION THREAD

**Discussion by Rush56** [arendst/Tasmota#22274] [Support request]
Hi.
This question is currently related to PID control driver. 
Most data in tasmota like template, rules and other parameters stay saved after reboot of device, but parameters like pidsp, pidpb, pidtd etc. get back to default values after rebooting. How to get these data to save after setting/changing them via console?

I noticed there is autoexec.bat option, which with weblog 4 was showing the file is not found. If i created autoexec.bat and i insert data, it will take these values after reboot, but i do not know how to make it to save changes in autoexec.bat after every changed input?

**Comment by sfromis** [comment]
When you have a file system, using `autoexec.bat` is indeed a good way of having a bunch of commands automatically executed after boot. Of course, you're right that this is a fixed set, not dynamic changes.
https://tasmota.github.io/docs/UFS/#autoexecbat

I suppose that the specific settings you mention were assumed to be pretty static, not being changed "all the time", and hence no feature for saving them automatically. It would be doable to create rules to pick up changes, save them in persistent `Mem` variables, and have a boot time rule reissuing the commands based on the saved variables.

**Reply by Rush56** [reply]
Okay thanks. Looking into it..

However, as you may already noticed from my previous posts i'm not to knowledgable with rules and functions.
I tryed with:
Rule1 on PID#Pidsp do mem1 %value% endon on System#boot do Pidsp mem1 %value% endon

Now the value is saved in mem, but after reboot it will pick up the default value again and save it as mem1. Did you mean another trigger instead system#boot or power#boot as boot time rule?

Could you give me an example or a link to similar setting maybe?

Thanks in advance..

**Reply by sfromis** [reply]
Dont use "mem1 %value%", but `%mem1%` in the boot rule, to have the content of the variable expanded - the "Pidsp" command don't know what the text "mem1" is. Other than that, looks ok.

**Reply by Rush56** [reply]
Rule1 on PID#Pidsp do mem1 %value% endon on power#boot do pidsp %mem1% endon 

Like that? It still picks the default value (19.5) after reboot. 

> 15:01:55.249 CMD: pidsp 50
> 15:01:55.256 RSL: RESULT = {"PidSp":50.0}
> 15:01:55.477 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:01:55.483 RSL: RESULT = {"Mem1":"50.00"}
> 15:01:56.694 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:01:56.700 RSL: RESULT = {"Mem1":"50.00"}
> 15:01:57.497 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:01:57.503 RSL: RESULT = {"Mem1":"50.00"}
> 15:01:58.476 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:01:58.482 RSL: RESULT = {"Mem1":"50.00"}
> 15:01:59.578 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:01:59.584 RSL: RESULT = {"Mem1":"50.00"}
> 15:02:00.581 RUL: PID#PIDSP performs 'mem1 50.00'
> 15:02:00.586 RSL: RESULT = {"Mem1":"50.00"}
> 
> 00:00:00.002 HDW: ESP8266EX
> 00:00:00.026 UFS: FlashFS mounted with 1992 kB free
> 00:00:00.093 CFG: Loaded from File, Count 135
> 00:00:00.099 QPC: Count 1
> 00:00:00.116 Project tasmota - Tasmota Version 14.2.0(TasmoCompiler-esp82664M)-2_7_7(2024-10-12T08:03:15)
> 00:00:01.001 WIF: Connecting to AP1 MikroTik Channel 6 BSSId 9C:A2:F4:C2:DE:4E in mode 11n as tasmota-5AC4E3-1251...
> 00:00:02.513 WIF: Connected
> 00:00:02.765 HTP: Web server active on tasmota-5AC4E3-1251 with IP address 192.168.2.6
> 15:02:12.215 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:12.219 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:12.507 RSL: INFO1 = {"Info1":{"Module":"Sonoff Basic","Version":"14.2.0(TasmoCompiler-esp82664M)","FallbackTopic":"cmnd/DVES_5AC4E3_fb/","GroupTopic":"cmnd/tasmotas/"}}
> 15:02:12.516 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-5AC4E3-1251","IPAddress":"192.168.2.6"}}
> 15:02:12.526 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":44}}
> 15:02:12.534 RSL: RESULT = {"POWER":"OFF"}
> 15:02:12.536 RSL: POWER = OFF
> 15:02:12.609 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:12.614 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:13.511 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:13.516 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:14.281 QPC: Reset
> 15:02:14.499 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:14.503 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:15.500 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:15.505 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:16.225 RSL: STATE = {"Time":"2024-10-12T15:02:16","Uptime":"0T00:00:09","UptimeSec":9,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"MikroTik","BSSId":"9C:A2:F4:C2:DE:4E","Channel":6,"Mode":"11n","RSSI":64,"Signal":-68,"LinkCount":1,"Downtime":"0T00:00:04"}}
> 15:02:16.254 RSL: SENSOR = {"Time":"2024-10-12T15:02:16","PID":{"PidPv":0.00,"PidSp":19.50,"PidShutdown":0,"PidPb":5.00,"PidTi":1800.00,"PidTd":15.00,"PidInitialInt":0.50,"PidDSmooth":3.00,"PidAuto":1,"PidManualPower":0.00,"PidMaxInterval":300,"PidInterval":9,"PidUpdateSecs":0,"PidSensorLost":0,"PidPower":0.00}}
> 15:02:16.719 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:16.725 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:17.522 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:17.527 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:18.526 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:18.531 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:19.529 RUL: PID#PIDSP performs 'mem1 19.50'
> 15:02:19.534 RSL: RESULT = {"Mem1":"19.50"}
> 15:02:20.590 RUL: PID#PIDSP performs 'mem1 1

Seems like the driver is loaded before applying the rule and changes mem1 back to default set..

**Reply by sfromis** [reply]
Not sure if you fixed the command as I suggested, or the rule is actually enabled.

Anyway, `Power#boot` is invalid, it should instead be for a specific channel, but `System#boot` will occur later than the correct one.

**Reply by Rush56** [reply]
Using Power1#boot it sets previous value at power up, but then gets back to default:

Rule1 on PID#Pidsp do mem1 %value% endon on Power1#boot do pidsp %mem1% endon

> 00:00:00.009 UFS: FlashFS mounted with 1992 kB free
> 00:00:00.070 CFG: Loaded from File, Count 196
> 00:00:00.075 QPC: Count 1
> 00:00:00.097 RUL: POWER1#BOOT performs 'pidsp 50.00'
> 00:00:00.102 RSL: RESULT = {"PidSp":50.0}
> 00:00:00.104 Project tasmota - Tasmota Version 14.2.0(TasmoCompiler-esp82664M)-2_7_7(2024-10-12T08:03:15)
> 00:00:01.001 WIF: Connecting to AP1 MikroTik Channel 6 BSSId 9C:A2:F4:C2:DE:4E in mode 11n as tasmota-5AC4E3-1251...
> 00:00:02.513 WIF: Connected
> 00:00:02.766 HTP: Web server active on tasmota-5AC4E3-1251 with IP address 192.168.2.6
> 16:09:37.223 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:37.228 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:37.514 RSL: INFO1 = {"Info1":{"Module":"Sonoff Basic","Version":"14.2.0(TasmoCompiler-esp82664M)","FallbackTopic":"cmnd/DVES_5AC4E3_fb/","GroupTopic":"cmnd/tasmotas/"}}
> 16:09:37.523 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-5AC4E3-1251","IPAddress":"192.168.2.6"}}
> 16:09:37.533 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":61}}
> 16:09:37.541 RSL: RESULT = {"POWER":"OFF"}
> 16:09:37.542 RSL: POWER = OFF
> 16:09:38.617 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:38.621 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:39.622 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:39.626 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:40.280 QPC: Reset
> 16:09:40.499 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:40.504 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:41.505 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:41.510 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:42.247 RSL: STATE = {"Time":"2024-10-12T16:09:42","Uptime":"0T00:00:10","UptimeSec":10,"Heap":23,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"MikroTik","BSSId":"9C:A2:F4:C2:DE:4E","Channel":6,"Mode":"11n","RSSI":60,"Signal":-70,"LinkCount":1,"Downtime":"0T00:00:04"}}
> 16:09:42.273 RSL: SENSOR = {"Time":"2024-10-12T16:09:42","PID":{"PidPv":0.00,"PidSp":19.50,"PidShutdown":0,"PidPb":5.00,"PidTi":1800.00,"PidTd":15.00,"PidInitialInt":0.50,"PidDSmooth":3.00,"PidAuto":1,"PidManualPower":0.00,"PidMaxInterval":300,"PidInterval":9,"PidUpdateSecs":0,"PidSensorLost":0,"PidPower":0.00}}
> 16:09:42.593 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:42.598 RSL: RESULT = {"Mem1":"19.50"}
> 16:09:43.496 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:09:43.501 RSL: RESULT = {"Mem1":"19.50"}

But if i use system#boot command it goes to default immediately.

Rule1 on PID#Pidsp do mem1 %value% endon on system#boot do pidsp %mem1% endon

> 00:00:00.001 HDW: ESP8266EX
> 00:00:00.012 UFS: FlashFS mounted with 1992 kB free
> 00:00:00.074 CFG: Loaded from File, Count 200
> 00:00:00.079 QPC: Count 1
> 00:00:00.093 Project tasmota - Tasmota Version 14.2.0(TasmoCompiler-esp82664M)-2_7_7(2024-10-12T08:03:15)
> 00:00:01.001 WIF: Connecting to AP1 MikroTik Channel 6 BSSId 9C:A2:F4:C2:DE:4E in mode 11n as tasmota-5AC4E3-1251...
> 00:00:02.514 WIF: Connected
> 00:00:02.767 HTP: Web server active on tasmota-5AC4E3-1251 with IP address 192.168.2.6
> 16:11:55.306 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:11:55.311 RSL: RESULT = {"Mem1":"19.50"}
> 16:11:55.469 RSL: INFO1 = {"Info1":{"Module":"Sonoff Basic","Version":"14.2.0(TasmoCompiler-esp82664M)","FallbackTopic":"cmnd/DVES_5AC4E3_fb/","GroupTopic":"cmnd/tasmotas/"}}
> 16:11:55.478 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-5AC4E3-1251","IPAddress":"192.168.2.6"}}
> 16:11:55.487 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":62}}
> 16:11:55.497 RSL: RESULT = {"POWER":"OFF"}
> 16:11:55.498 RSL: POWER = OFF
> 16:11:55.531 RUL: SYSTEM#BOOT performs 'pidsp 19.50'
> 16:11:55.536 RSL: RESULT = {"PidSp":19.5}
> 16:11:56.682 RUL: PID#PIDSP performs 'mem1 19.50'
> 16:11:56.686 RSL: RESULT = {"Mem1":"19.50"}
> 16:11:57.587 RUL: PID#PIDSP performs 'mem1 19.5

**Comment by Rush56** [comment]
I managed to get it working with this command:

Rule1 on PID#Pidsp do mem1 %value% endon on Power1#boot do backlog pidsp %mem1% endon
The "backlog" was missing in previous rule.

edit: Is there a way to make it perform action only on PID#Pidsp value change and not every second? Floods the console....

Thank you very much for your help i really appreciate it. Have a lot of hobby projects with automations but still need to learn a lot :)

**Reply by sfromis** [reply]
Well, you can have your rule only act for `PID#Pidsp!=%mem1%` instead of any update (changed or not). 

BTW, using `Backlog` in the rule should make no difference, except to the immediate action of the rule set picking up the value. The problem was using `Power` instead of `Power1`

**Reply by Rush56** [reply]
Yes i made a mistake with Power instead of Power1 before. But i used Power1 later on.
However i mistakenly added backlog as i was changing and adding commands to rule - wanted to turn rule off after boot, then it started to work properly. I thought the solution was when value was taken from mem1 and after that turned rule off, before defaults were loaded. However i then removed command to turn rule off and left backlog there and it was working properly.

Then to confirm i removed backlog and tryed again and it did not work, it took default values. Adding backlog back there and it works properly now. 

I dont know why and how, but thats how it worked out....

You can see in the last post with console quote, i copy&pasted the rule and is written correctly, yet you can see in console response it went back to defaults.


> Well, you can have your rule only act for PID#Pidsp!=%mem1% instead of any update (changed or not).

Awesome.. Took me a while to figure out how to use this info and yeah, that's exactly what i was looking for. 

Thanks again many times!
